### PR TITLE
Infer rank where reshape shape is inferred

### PR DESCRIPTION
**Pull request by adityagoel4512**
Description
This PR should address https://github.com/onnx/onnx/issues/5324. In situations where the shape in a dynamic reshape operation itself has an inferred shape, this can be used to determine the rank of the output shape.

Motivation and Context
Closes https://github.com/onnx/onnx/issues/5324

### CODE REVIEW

File: onnx/defs/tensor/defs.cc
Comment (gramalingam) [inline onnx/defs/tensor/defs.cc:321]: Thanks for the improvement! I think we can simplify the whole inference by calling [getShapeInput](https://github.com/onnx/onnx/blob/3dafb54b11a496cd933de17fd303324d0686fc3b/onnx/defs/shape_inference.cc#L449) which combines these three modes of getting shape information, including the new one below.

Comment (adityagoel4512) [reply onnx/defs/tensor/defs.cc]: Thanks for the tip! That simplifies things quite a lot.

File: onnx/test/shape_inference_test.py
Comment (adityagoel4512) [inline onnx/test/shape_inference_test.py:70]: Just a quick note on this. getShapeInput's rank inference [fails hard](https://github.com/onnx/onnx/blob/3dafb54b11a496cd933de17fd303324d0686fc3b/onnx/defs/shape_inference.cc#L475) when the shape is not 1 dimensional which makes this necessary. I would be in favour of the function setting `found` to `false` and returning instead since this lets the caller determine how to proceed.

Comment (adityagoel4512) [inline onnx/test/shape_inference_test.py:70]: Just a quick note on this. getShapeInput's rank inference [fails hard](https://github.com/onnx/onnx/blob/3dafb54b11a496cd933de17fd303324d0686fc3b/onnx/defs/shape_inference.cc#L475) when the shape is not 1 dimensional which makes this necessary.

Comment (xadupre) [reply onnx/test/shape_inference_test.py:70]: Maybe it would be worth modifying getShapeInput to avoid failing. Some operators takes scalar as inputs. Can it return something in case there is no dimension?

Comment (adityagoel4512) [reply onnx/test/shape_inference_test.py:70]: Yes, I think that's quite reasonable. I had a more general question: is it clearly defined when we should `fail_shape_inference` and when we should simply not propagate shape information?

Comment (xadupre) [reply onnx/test/shape_inference_test.py]: @gramalingam would know better. If it does not fail, we could probably retrieve shape inference for the first nodes. Would users be interested into partial shape inference? That may be an interesting feature for another PR.

Comment (gramalingam) [reply onnx/test/shape_inference_test.py]: It is common for inference to throw a shape-inference-error if an input is detected as having the wrong rank or shape. (The outer inference loop collects all such error messages, and tries to continue with inference for all nodes.)

I think it is okay for Reshape's shape-inference to fail if the input is detected as having the wrong rank. Are you running into any specific issue because of this?

Comment (gramalingam) [reply onnx/test/shape_inference_test.py]: As far as I can see, the change you proposed (in old line 70, now moved to different line due to merged PR) to specify UNKNOWN_SHAPE's shape as `(None,)` seems correct. The older `()` seems incorrect.

Comment (adityagoel4512) [reply onnx/test/shape_inference_test.py:111]: Well the only issue was that UNKNOWN_SHAPE was `()` but I think `(None,)` is what it should be. Given that, I'll let Reshape continue to fail hard if shape has an invalid rank. It should be ready for another look @gramalingam

Comment (gramalingam) [reply onnx/test/shape_inference_test.py:111]: LGTM, thanks!

**Review by gramalingam — APPROVED** [review]

**Comment by adityagoel4512** [comment]
Does this require another approval?

**Comment by gramalingam** [comment]
> Does this require another approval?

No. I have enabled auto-merge, so it should merge once CI completes, unless some other PR merges first (forcing another round of merging).

**Pull request merged**
